aea/helpers/profiling.py

```diff
@@ -0,0 +1,162 @@
+        return 1.0 * d["WorkingSetSize"] / 1024 ** 2
+
+    def get_current_process_cpu_time() -> float:
+        """Get current process cpu time in seconds."""
+        d = win32process.GetProcessTimes(win32process.GetCurrentProcess())  # type: ignore
+        return d["UserTime"] / WIN32_PROCESS_TIMES_TICKS_PER_SECOND
+
+
+else:
+    import resource
+
+    if platform.system() == "Darwin":  # pragma: nocover
+        divider = 1024 ** 2
+    else:
+        divider = 1024
+
+    def get_current_process_memory_usage() -> float:
+        """Get current process memory usage in MB."""
+        return 1.0 * resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / divider
+
+    def get_current_process_cpu_time() -> float:
+        """Get current process cpu time in seconds."""
+        return resource.getrusage(resource.RUSAGE_SELF).ru_utime
+
+
+class Profiling(Runnable):
+    """Profiling service."""
+
+    OBJECTS = [Message, Dialogue, Handler, Behaviour, Connection]
+
+    def __init__(
+        self,
+        period: int = 0,
+        output_function: Callable[[str], None] = lambda x: print(x, flush=True),
+    ) -> None:
+        """
+        Init profiler.
+
+        :param period: delay between profiling output in seconds.
+        :param output_function: function to display ouput, one str argument.
+        """
+        if period < 1:  # pragma: nocover
+            raise ValueError("Period should be at least 1 second!")
+        super().__init__(threaded=True)
+        self._period = period
+        self._start_ts = time.time()
+        self._messages_count = 0
+        self._output_function = output_function
+
+    def set_messages_counter(self) -> None:
+        """Modify Message.__new__ to count messages created."""
+        orig_new = Message.__new__  # pylint: disable=protected-access  # type: ignore
+
+        def new(*args, **kwargs) -> Message:
+            self._messages_count += 1
+            return orig_new(*args, **kwargs)
+
+        Message.__new__ = new  # type: ignore
+
+    async def run(self) -> None:
+        """Run profiling."""
+        try:
+            self.set_messages_counter()
+            while True:
+                await asyncio.sleep(self._period)
+                await self.output_profile_data()
+        except Exception:  # pragma: nocover
+            logger.exception("Exception in Profiling")
+            raise
+
+    async def output_profile_data(self) -> None:
+        """Render profiling data and call output_function."""
+        data = self.get_profile_data()
+        text = (
+            textwrap.dedent(
+                f"""
+        Profiling details: {datetime.datetime.now()}
+        =============================================
+        Run time: {data["run_time"]} seconds
+        Cpu time: {data["cpu_time"]} seconds
+        Memory: {data["mem"]} MB
+        Threads: {data["threads"]}
+        Messages constructed: {data["messages_contstructed"]}
```

Comment:
Dialogues constructed should be shown too.

Reply:
done